Conform `Array` to various `ExpressibleAs` protocols in `BuildableCollectionNodes`

fwcd opened the pull request:
In a separate PR (from #460) for easier reviewability. This adds a bunch of `Array` conformances that let us initialize syntax collections more conveniently, for example:

```swift
ExtensionDecl(
  modifiers: [...],
  ...
)
```

instead of

```swift
ExtensionDecl(
  modifiers: ModifierList([...]),
  ...
)
```

(This applies whenever the parameter has an existential type a la `ExpressibleAsModifierList`, in which case the compiler cannot infer that we would like to use the `ExpressibleByArrayLiteral` conformance from `ModifierList`)

cc @ahoppen

kimdv commented:
@swift-ci please test

ahoppen reviewed (commented):
Nice. This looks great. Could you add a couple test cases?

fwcd commented:
Sure, I've added a few tests. Additionally, I have updated the conformance clause to always test for equality against the element type, i.e. use existentials. This is slightly more flexible and more consistent with the existing initializers e.g. in `ModifierList`, which also take a list of existentials (`[ExpressibleAsDeclModifier]`).

kimdv commented:
@swift-ci please test

ahoppen reviewed (approved):
Thanks. Looks great!
